File: v2ProjectionMatrix/cube.py
from settings import *

# still not perfect, come back and modify


class Cube:
    def __init__(self, game):
        self.game = game
        self.matrixA = [[-1, 1, -3],
                        [1, 1, -3],
                        [1, -1, -3],
                        [-1, -1, -3],  # --
                        [-1, -1, -5],
                        [1, -1, -5],
                        [1, 1, -5],
                        [-1, 1, -5]]
        self.matrixAt = [[0, 0, 0],
                         [0, 0, 0],
                         [0, 0, 0],
                         [0, 0, 0],
                         [0, 0, 0],
                         [0, 0, 0],
                         [0, 0, 0],
                         [0, 0, 0]]
        self.matrixB = [[-1, -1, 1],
                        [1, -1, 1],
                        [1, 1, 1],
                        [-1, 1, 1],
                        [-1, -1, -1],
                        [1, -1, -1],
                        [1, 1, -1],
                        [-1, 1, -1]]
        self.th = 0
        self.dist = 3

    def draw2(self):
        self.th += 0.01
        mRx = [[1, 0, 0],
               [0, math.cos(self.th), math.sin(self.th)],
               [0, -math.sin(self.th), math.cos(self.th)]]
        mRy = [[math.cos(self.th), 0, -math.sin(self.th)],
               [0, 1, 0],
               [math.sin(self.th), 0, math.cos(self.th)]]
        mRz = [[math.cos(self.th), math.sin(self.th), 0],
               [-math.sin(self.th), math.cos(self.th), 0],
               [0, 0, 1]]

        i = 0
        for p in self.matrixB:
            RxPx = p[0] * mRx[0][0] + p[1] * mRx[1][0] + p[2] * mRx[2][0]
            RxPy = p[0] * mRx[0][1] + p[1] * mRx[1][1] + p[2] * mRx[2][1]
            RxPz = p[0] * mRx[0][2] + p[1] * mRx[1][2] + p[2] * mRx[2][2]

            RyPx = RxPx * mRy[0][0] + RxPy * mRy[1][0] + RxPz * mRy[2][0]
            RyPy = RxPx * mRy[0][1] + RxPy * mRy[1][1] + RxPz * mRy[2][1]
            RyPz = RxPx * mRy[0][2] + RxPy * mRy[1][2] + RxPz * mRy[2][2]

            RzPx = RyPx * mRz[0][0] + RyPy * mRz[1][0] + RyPz * mRz[2][0]
            RzPy = RyPx * mRz[0][1] + RyPy * mRz[1][1] + RyPz * mRz[2][1]
            RzPz = RyPx * mRz[0][2] + RyPy * mRz[1][2] + RyPz * mRz[2][2]

            z = 1 / (self.dist - RzPz)  # t1 scales it down and makes larger dist further instead of reverse
            # z = 1  # uncomment for orthographic projection

            pjM = [[z, 0, 0],
                   [0, z, 0]]

            x = RzPx * pjM[0][0] + RzPy * pjM[1][0]
            y = RzPx * pjM[0][1] + RzPy * pjM[1][1]
            x = int(x * SCALE + HWIDTH)
            y = int(y * SCALE + HHEIGHT)
            pg.draw.circle(self.game.screen, 'red', (x, y), 5)
            print(x)

            self.matrixAt[i][0] = x
            self.matrixAt[i][1] = y
            i += 1
        self.draw_lines()

    def draw_lines(self):
        for i in range(4):
            self.connect(i, (i + 1) % 4)
            self.connect(i + 4, (i + 1) % 4 + 4)
            self.connect(i, i + 4)
            # new concept: using modulus

    def connect(self, i, j):
        pg.draw.line(self.game.screen, 'red', (self.matrixAt[i][0], self.matrixAt[i][1]),
                     (self.matrixAt[j][0], self.matrixAt[j][1]))

File: v3ProjMatrixCpp/cube.cpp
#include "cube.h"

Cube::Cube() {
    pt.radius = 5;
}

void Cube::Update() {
    th += 0.01;
    vector<Vector3> mRx = {
        Vector3{1, 0, 0},
        Vector3{0, cos(th), sin(th)},
        Vector3{0, -sin(th), cos(th)} };

    vector<Vector3> mRy = {
        Vector3{cos(th), 0, -sin(th)},
        Vector3{0, 1, 0},
        Vector3{sin(th), 0, cos(th)} };

    vector<Vector3> mRz = {
        Vector3{cos(th), sin(th), 0},
        Vector3{-sin(th), cos(th), 0},
        Vector3{0, 0, 1} };

    for (int i = 0; i < matrixA.size(); i++) {
        Vector3 p = matrixA[i];
        float RxPx = p.x * mRx[0].x + p.y * mRx[1].x + p.z * mRx[2].x;
        float RxPy = p.x * mRx[0].y + p.y * mRx[1].y + p.z * mRx[2].y;
        float RxPz = p.x * mRx[0].z + p.y * mRx[1].z + p.z * mRx[2].z;

        float RyPx = RxPx * mRy[0].x + RxPy * mRy[1].x + RxPz * mRy[2].x;
        float RyPy = RxPx * mRy[0].y + RxPy * mRy[1].y + RxPz * mRy[2].y;
        float RyPz = RxPx * mRy[0].z + RxPy * mRy[1].z + RxPz * mRy[2].z;

        float RzPx = RyPx * mRz[0].x + RyPy * mRz[1].x + RyPz * mRz[2].x;
        float RzPy = RyPx * mRz[0].y + RyPy * mRz[1].y + RyPz * mRz[2].y;
        float RzPz = RyPx * mRz[0].z + RyPy * mRz[1].z + RyPz * mRz[2].z;

        float z = 1 / (dist - RzPz);

        vector<Vector3> pjM = {
            Vector3{z, 0, 0},
            Vector3{0, z, 0}
        };

        float pX = RzPx * pjM[0].x + RzPy * pjM[1].x;
        float pY = RzPx * pjM[0].y + RzPy * pjM[1].y;

        pX = int(pX * st.scale + st.scrWid / 2);
        pY = int(pY * st.scale + st.scrHei / 2);

        DrawCircle(pX, pY, pt.radius, WHITE);

        matrixAt[i].x = pX;
        matrixAt[i].y = pY;
    }
    ConnectLine();
}

void Cube::ConnectLine() {
    for (int i = 0; i < 4; i++) {
        Connect(i, (i + 1) % 4);
        Connect(i + 4, (i + 1) % 4 + 4);
        Connect(i, i + 4);
    }
}

void Cube::Connect(int i, int j) {
    DrawLine(matrixAt[i].x, matrixAt[i].y, matrixAt[j].x, matrixAt[j].y, WHITE);
}